pkg/engine/executor/sortmerge.go

```diff
@@ -0,0 +1,202 @@
+		compare = func(a, b uint64) bool { return a >= b }
+	default:
+		return nil, fmt.Errorf("invalid sort order %v", order)
+	}
+
+	return &KWayMerge{
+		inputs:     inputs,
+		columnEval: evaluator.newFunc(column),
+		compare:    compare,
+	}, nil
+}
+
+// KWayMerge is a k-way merge of multiple sorted inputs.
+// It requires the input batches to be sorted in the same order (ASC/DESC) as the SortMerge operator itself.
+// The sort order is defined by the direction of the query, which is either FORWARD or BACKWARDS,
+// which is applied to the SortMerge as well as to the DataObjScan during query planning.
+type KWayMerge struct {
+	inputs      []Pipeline
+	state       state
+	initialized bool
+	batches     []arrow.Record
+	exhausted   []bool
+	offsets     []int64
+	columnEval  evalFunc
+	compare     func(a, b uint64) bool
+}
+
+var _ Pipeline = (*KWayMerge)(nil)
+
+// Close implements Pipeline.
+func (p *KWayMerge) Close() {
+	// Release last batch
+	if p.state.batch != nil {
+		p.state.batch.Release()
+	}
+	for _, input := range p.inputs {
+		input.Close()
+	}
+}
+
+// Inputs implements Pipeline.
+func (p *KWayMerge) Inputs() []Pipeline {
+	return p.inputs
+}
+
+// Read implements Pipeline.
+func (p *KWayMerge) Read() error {
+	p.init()
+	return p.read()
+}
+
+// Transport implements Pipeline.
+func (p *KWayMerge) Transport() Transport {
+	return Local
+}
+
+// Value implements Pipeline.
+func (p *KWayMerge) Value() (arrow.Record, error) {
+	return p.state.Value()
+}
+
+func (p *KWayMerge) init() {
+	if p.initialized {
+		return
+	}
+
+	p.initialized = true
+
+	n := len(p.inputs)
+	p.batches = make([]arrow.Record, n)
+	p.exhausted = make([]bool, n)
+	p.offsets = make([]int64, n)
+
+	if p.compare == nil {
+		p.compare = func(a, b uint64) bool { return a <= b }
+	}
+}
+
+// Iterate through each record, looking at the value from their starting slice offset.
+// Track the top two winners (e.g., the record whose next value is the smallest and the record whose next value is the next smallest).
+// Find the largest offset in the starting record whose value is still less than the value of the runner-up record from the previous step.
+// Return the slice of that record using the two offsets, and update the stored offset of the returned record for the next call to Read.
+func (p *KWayMerge) read() error {
+	// Release previous batch
+	if p.state.batch != nil {
+		p.state.batch.Release()
+	}
+
+	timestamps := make([]uint64, 0, len(p.inputs))
+	batchIndexes := make([]int, 0, len(p.inputs))
+
+	for i := range len(p.inputs) {
+		// Skip exhausted inputs
+		if p.exhausted[i] {
+			continue
+		}
+
+		// Load next batch if it hasn't been loaded yet, or if current one is already fully consumed
+		if p.batches[i] == nil || p.offsets[i] == p.batches[i].NumRows() {
+			err := p.inputs[i].Read()
+			if err != nil {
+				if err == EOF {
+					p.exhausted[i] = true
+					continue
+				}
+				return err
+			}
+			p.offsets[i] = 0
+			p.batches[i], _ = p.inputs[i].Value()
```

Comment:
any chance of this returning non-nil? (If not, adding a quick comment that it's not possible and why would be helpful)

Reply:
Will add a comment. The reason we cannot have a nil value is because we already check the error from the `Read()` call.